ts/models/models.ts

```diff
@@ -0,0 +1,45 @@
+import { assign, cloneDeep } from 'lodash';
+import { MessageModel } from './message';
+import type { MessageAttributes } from './messageType';
+
+export type ModelAttributes = Record<string, any> & { id: string };
+
+export abstract class Model<T extends ModelAttributes> {
+  protected _attributes: T;
+
+  constructor(attributes: T) {
+    this._attributes = attributes;
+  }
+
+  get id(): string {
+    return this._attributes.id;
+  }
+
+  public get<K extends keyof T>(key: K): T[K] {
+    return this._attributes[key];
+  }
+
+  public setKey<K extends keyof T>(key: K, value: T[K] | undefined) {
+    const toSet: Partial<T> = {};
+    toSet[key] = value;
+    this.set(toSet);
+    return this;
+  }
+
+  public set(attrs: Partial<T>) {
+    this._attributes = assign(this._attributes, attrs);
+    return this;
+  }
+
+  get attributes(): Readonly<T> {
+    return this._attributes;
+  }
+
+  public cloneAttributes() {
+    return cloneDeep(this._attributes);
+  }
+}
+
+export function makeMessageModels(modelsOrAttrs: Array<MessageModel | MessageAttributes>) {
+  return modelsOrAttrs.map(a => (a instanceof MessageModel ? a : new MessageModel(a)));
```

Comment:
Why would you pass models directly here?

Reply:
we use this with a lot of untyped code (ipc calls) that are `any`, I figured having this check here would prevent an issue down the line, but happy to remove it.

Reply:
All good to keep it lets just have a TODO to remove in future once our typings are more robust